### ISSUE TITLE
Is Symfony process component work different between Linux and Max OS X?

### ISSUE THREAD

**Issue by TonyGao** [symfony/symfony#20445]
I got different result in the same code like below

With a **Symfony command A** I create a process running at background asynchronously, and it work fine in Linux, and I got the new process and the Pid as the command output.

**Symfony command A**
```php
$command = 'php /somePath/app/console someCommand:processA';
$process = new Process($command);
$process->start();

$output->writeln($process->getPid());
```

**Symfony command processA**
```php
$number = 1;
while (true) {
    $number++;
    $output->writeln($number);
}
```

But I could not got the process in Mac OS X, except I keep the **Symfony command A** live. Is there something wrong with the code or something different between Linux and Mac?

Linux: Ubuntu 15.10, PHP 5.6.11, Symfony 3.1.6
Mac OS: 10.11.6, PHP 5.5.38, Symfony 2.8.12

**Comment by jakzal** [comment]
The goal of this component is to provide consistent process management across popular operating systems.

What happens on OS X exactly?  I'm not sure what you mean by " I could not got the process in Mac OS X". Is the PID not returned, or no output shown from the process?

By the way, if you run symfony commands you need to be aware of: http://symfony.com/doc/current/components/process.html#process-pid

**Comment by TonyGao** [comment]
@jakzal Exactly in OS X, $process->start() just like crash,  Nothing happen and there's no any php process in the process moniter. For instance by what I'm doing now

```
$process = new Process("exec afplay $pathToMusic");
$process->setTimeout(null);
$process->run();
$output->writeln('Music End baby :D');
```

afplay is a application to play music file in console, everything works well above, even without exec. Music show is going on, at the end I can see the ending words. But if I change run() method to start(), I just see "Music End baby :D" but without enjoying music anymore :( 

In Linux start() method means keep working like a daemon, but I find there is no same effect in OS X . Any idea?

**Comment by jakzal** [comment]
This is an expected behaviour.

Process you started with `start()` will die as soon as the parent process finishes. It will be stopped from [the process'es destructor](https://github.com/symfony/process/blob/master/Process.php#L179). Otherwise it would be terminated by the operating system. When using `start()` you need to use either [`wait()` or `isProcessRunning()`](http://symfony.com/doc/current/components/process.html#running-processes-asynchronously) to make sure you're not finishing the parent process before the child finishes. 

Not sure why the linux process doesn't terminate in your case.

**Comment by stof** [comment]
@jakzal this is because on Linux, killing a process does not kill its child processes. So only the sh wrapper gets killed, and not the actual command on Linux.

**Comment by tristanbes** [comment]
@jakzal wait, what ? According to you, we can't use the `start` method to say "ok launch this heavy task (eg: symfony command) and forget about it ?` What's the point of running an asynchronous process if the subprocess is killed when the parent is kiled (basically when you return a response for example ?)

How to achieve the fire and forget thing with the `Process` component so ? running it async and having to do `isProcessRunning()` kind of kill the benefit of the async feature :(

**Comment by stof** [comment]
running an asynchronous process is not the same than running a processing surviving yourselves.
the goal of an asynchronous process is to be able to do other thing in your master process before dealing with the output of the subprocess (a typical case being to launch several subprocesses in parallel)

**Comment by TonyGao** [comment]
@stof got it, but I even use it as a feature in Linux, because I found it is really similar to a process of fork. I'm not sure if I can use $process = new Process("$consoleCommand &"); to make the process work in the background itself ? Is that suggested? Or something other solution like pcntl?

**Comment by nicolas-grekas** [comment]
This is unsupported on Symfony Process because the behavior of the OSes is different. There is no way I'm aware about to abstract these differences unfortunately. It doesn't mean you can't achieve it on your setup, it just mean nobody wrote such portable code for Symfony (and I doubt PHP provides the required primitives).
For this reason, running a processing surviving yourselves is a won't fix for me unfortunately.

**Comment by jakzal** [comment]
As a result of #20513 [a docs issue was raised](https://github.com/symfony/symfony-docs/issues/7151) to clarify this behaviour in the documentation.
